### PR TITLE
refactor(chart): migrate center label plugin

### PR DESCRIPTION
**Pull request by Theeraphat-Sorasetsakul**
## Description

migrate centerLabel plugin of donut chart to compatible with ChartJS v4.

Fixes # ([STG-199](https://jira.refinitiv.com/browse/STG-199))

## Type of change

- [x] Refactor (improves code without changing its functionality)
- [x] This change requires a documentation update

### CODE REVIEW

File: packages/elements/src/chart/index.ts [hunk 1]
```diff
@@ -36,6 +36,10 @@ import type { ChartConfiguration, ChartOptions, UpdateMode, LegendItem } from 'c
 import type { Header } from '../header';
 import '../header/index.js';
 
+// Register plugins
+import doughnutCenterPlugin from './plugins/doughnut-center-label.js';
+ChartJS.register(doughnutCenterPlugin);
```
Comment (wattachai-lseg) [inline packages/elements/src/chart/index.ts]: instead of registering the plugin globally, let's add this plugin to `this.config` so that it would be registered to each chart instance.

File: packages/elements/src/chart/plugins/doughnut-center-label.ts [hunk 2]
```diff
@@ -1,257 +1,275 @@
...
+
+interface CenterLabelConfig {
+  defaultText: CenterLabel[];
+  selected: Selected;
+  onRenderLabel(chart: ChartJS, chartItems: ActiveElement[]): CenterLabel[];
+}
+
+declare module 'chart.js' {
+  interface PluginOptionsByType<TType extends ChartType> {
+    'centerLabel'?: CenterLabelConfig
```
Comment (wattachai-lseg) [inline packages/elements/src/chart/plugins/doughnut-center-label.ts:39]: `center-label` per chart.js plugin naming convention?

File: packages/elements/src/chart/plugins/doughnut-center-label.ts [hunk 3]
```diff
@@ -1,257 +1,275 @@
...
+ * @param {MetaData[]} active active data
+ * @returns {void}
+ */
+const drawItemBorder = function (chart: ChartJS, active: ActiveElement[]): void {
+
+  if (!chart.data.datasets) {
+    return;
+  }
+
+  if (active?.length) {
```
Comment (wattachai-lseg) [inline packages/elements/src/chart/plugins/doughnut-center-label.ts]: let's refactor to
```javascript
if (!active?.length) {
  return
}
```

File: packages/elements/src/chart/plugins/doughnut-center-label.ts [hunk 4]
```diff
@@ -1,257 +1,275 @@
...
+    const ctx = chart.ctx;
+    const chartItem = active[0];
+    const element = chartItem.element as ArcElement;
+    const vm = chartItem.element.options as ArcOptions;
```
Comment (wattachai-lseg) [inline packages/elements/src/chart/plugins/doughnut-center-label.ts]: rename `vm` to `archOptions`?

File: packages/elements/src/chart/plugins/doughnut-center-label.ts [hunk 5]
```diff
@@ -1,257 +1,275 @@
...
+    const datasets = chart.data.datasets[chartItem.datasetIndex];
+
+    if (Array.isArray(datasets.backgroundColor)) {
+      vm.backgroundColor = datasets.backgroundColor[chartItem.index] as string;
+      const color = parseColor(vm.backgroundColor);
+      if (color) {
+        vm.backgroundColor = color.darker(0.5).toString();
```
Comment (wattachai-lseg) [inline packages/elements/src/chart/plugins/doughnut-center-label.ts]: just to double check, this `darker(0.5)` bahaves the same way with the old `getHoverColor()`, right?

File: packages/elements/src/chart/plugins/doughnut-center-label.ts [hunk 6]
```diff
@@ -1,257 +1,275 @@
...
+      }
+    }
+    // vm.backgroundColor = CHART.helpers.getHoverColor(vm.backgroundColor); // we need to make color bolder
```
Comment (wattachai-lseg) [inline packages/elements/src/chart/plugins/doughnut-center-label.ts]: remove?

File: packages/elements/src/chart/plugins/doughnut-center-label.ts [hunk 7]
```diff
@@ -1,257 +1,275 @@
...
+
+    if (!config) {
+      return;
+    }
+
+    // Set render hook function
+    let active:ActiveElement[] = chart.getActiveElements();
+    if (chart._select) {
+      active = chart._select;
+    }
```
Comment (wattachai-lseg) [inline packages/elements/src/chart/plugins/doughnut-center-label.ts]:
```suggestion
    let active = chart._select ? chart._select : chart.getActiveElements();
```

File: packages/elements/src/chart/plugins/doughnut-center-label.ts [hunk 8]
```diff
@@ -1,257 +1,275 @@
...
+    const defaultFontStyle = globalConfig.font.family;
+
+    const defaultColor = style.getPropertyValue('--doughnut-center-text-color') || globalConfig.color;
+    const backgroundColor = style.getPropertyValue('--doughnut-center-background-color');
+    const customFontSizePercent = parseInt(style.getPropertyValue('--doughnut-center-font-size'), 10) / 100; // font size in percentage
+
+    const centerX = (chart.chartArea.left + chart.chartArea.right) / 2;
+    const centerY = (chart.chartArea.top + chart.chartArea.bottom) / 2;
+
+    const innerRadius = (chart.getDatasetMeta(chart.data.datasets.length - 1).data[0] as unknown as DoughnutController).innerRadius;
```
Comment (wattachai-lseg) [inline packages/elements/src/chart/plugins/doughnut-center-label.ts:182]: why are we using the last data of dataset?

File: packages/elements/src/chart/plugins/doughnut-center-label.ts [hunk 9]
```diff
@@ -1,257 +1,275 @@
...
+    }
+
+    // Set selected item to the chart
+    const items = chart.getDatasetMeta(datasetIndex).data[selectedIndex];
+    chart.selected = items ? [{ element: items as ArcElement, datasetIndex, index: selectedIndex }] : [];
+  },
+  afterDatasetsDraw: function (chart: ChartJS & Selectable): void {
+    if (getPluginConfig(chart)) {
+      // Draw active element
+      // Note: use logic from chart.js - chart.js/src/elements/element.arc.js :draw()
```
Comment (wattachai-lseg) [inline packages/elements/src/chart/plugins/doughnut-center-label.ts]: do we still need this comment? It looks irrelevant.

File: packages/elements/src/chart/__demo__/index.html [hunk 10]
```diff
@@ -411,29 +412,32 @@
...
-                var title = data.labels[tooltipItem.index];
-                var result = data.datasets[0].data[tooltipItem.index];
-                return title + ': ' + result + '%';
+          plugins: {
+            title: {
+              text: 'Operating Segments of AAPL.O in 2014'
+            },
+            tooltip: {
+              callbacks: {
+                title : () => null,
```
Comment (wattachai-lseg) [inline packages/elements/src/chart/__demo__/index.html]: eslint issue: let's be consistent.

File: packages/elements/src/chart/__demo__/index.html [hunk 11]
```diff
@@ -495,12 +505,7 @@
               }
             }
           },
```
Comment (wattachai-lseg) [inline packages/elements/src/chart/__demo__/index.html]: eslint: trailing comma

**Review by wattachai-lseg — CHANGES_REQUESTED** [review]
Notice that center label plugin bahaves a little different from EFv6:
* hovering each segment of the chart doesn't update center label
* tooltip doesn't behave as before
  * v6: `${segment name}: ${value}`
  * v7: `${segment name}\n${value}`

File: packages/elements/src/chart/plugins/doughnut-center-label.ts (hunk 2 above)
Comment (Theeraphat-Sorasetsakul) [reply packages/elements/src/chart/plugins/doughnut-center-label.ts:39]: As we discussed, we aren't going to change. Because if we change to be kebab-case, it will cause breaking change that users have to put id plugin as key string. It will be improper to declare object like below:
```javascript
options: {
 plugins: {
   'center-label': {
     defaultText: ...,
     ...
   }
 }
}
```

File: packages/elements/src/chart/__demo__/index.html (hunk 11 above)
Comment (Theeraphat-Sorasetsakul) [reply packages/elements/src/chart/__demo__/index.html]: Fixed.

File: packages/elements/src/chart/__demo__/index.html (hunk 10 above)
Comment (Theeraphat-Sorasetsakul) [reply packages/elements/src/chart/__demo__/index.html]: Fixed.

File: packages/elements/src/chart/plugins/doughnut-center-label.ts (hunk 9 above)
Comment (Theeraphat-Sorasetsakul) [reply packages/elements/src/chart/plugins/doughnut-center-label.ts]: Removed.

File: packages/elements/src/chart/index.ts (hunk 1 above)
Comment (Theeraphat-Sorasetsakul) [reply packages/elements/src/chart/index.ts]: Updated.

File: packages/elements/src/chart/plugins/doughnut-center-label.ts (hunk 3 above)
Comment (Theeraphat-Sorasetsakul) [reply packages/elements/src/chart/plugins/doughnut-center-label.ts]: Updated.

File: packages/elements/src/chart/plugins/doughnut-center-label.ts (hunk 6 above)
Comment (Theeraphat-Sorasetsakul) [reply packages/elements/src/chart/plugins/doughnut-center-label.ts]: Removed.

File: packages/elements/src/chart/plugins/doughnut-center-label.ts (hunk 4 above)
Comment (Theeraphat-Sorasetsakul) [reply packages/elements/src/chart/plugins/doughnut-center-label.ts]: Updated.

File: packages/elements/src/chart/plugins/doughnut-center-label.ts (hunk 5 above)
Comment (Theeraphat-Sorasetsakul) [reply packages/elements/src/chart/plugins/doughnut-center-label.ts]: I just did wrong implementation. So, I updated to use getHoverColor from 'chart.js/helpers'.

File: packages/elements/src/chart/plugins/doughnut-center-label.ts (hunk 7 above)
Comment (Theeraphat-Sorasetsakul) [reply packages/elements/src/chart/plugins/doughnut-center-label.ts]: I added logic to fix hovering each datasets. Please take a look my update.

File: packages/elements/src/chart/plugins/doughnut-center-label.ts (hunk 8 above)
Comment (Theeraphat-Sorasetsakul) [reply packages/elements/src/chart/plugins/doughnut-center-label.ts:182]: innerRadius refers to center circle hole radius. That the lastest dataset is inmost datasets.

File: packages/elements/src/chart/plugins/doughnut-center-label.ts (hunk 9 above)
Comment (wattachai-lseg) [reply packages/elements/src/chart/plugins/doughnut-center-label.ts]: on a second thought, we should make a small adjustment to the comment and move it to `drawItemBorder()` instead.

File: packages/elements/src/chart/plugins/doughnut-center-label.ts (hunk 8 above)
Comment (wattachai-lseg) [reply packages/elements/src/chart/plugins/doughnut-center-label.ts:182]: let's rephase the comment to "radius of circular hole in the middle of the chart"

File: packages/elements/src/chart/plugins/doughnut-center-label.ts (hunk 7 above)
Comment (wattachai-lseg) [reply packages/elements/src/chart/plugins/doughnut-center-label.ts]: the new one doesn't look right. Let's discuss more.

File: packages/elements/src/chart/plugins/doughnut-center-label.ts (hunk 9 above)
Comment (Theeraphat-Sorasetsakul) [reply packages/elements/src/chart/plugins/doughnut-center-label.ts]: Moved to upper drawItemBorder function.

File: packages/elements/src/chart/plugins/doughnut-center-label.ts (hunk 8 above)
Comment (Theeraphat-Sorasetsakul) [reply packages/elements/src/chart/plugins/doughnut-center-label.ts:182]: Updated.

**Review by wattachai-lseg — APPROVED** [review]
LGTM
